Django Backend Re-write

davidswarbrick opened the pull request:
After the discussion of #49,  many of the features this project may require going forwards such as
-  An administrator/moderator login page, 
-  Fixed models for data stored in the project's database
-  An understandable & well documented information flow from database to HTML
-  Extendability to a REST framework

These are all already features of Django, or can be easily added through extension while retaining a well-defined information flow and representation. Although the Flask backend is coping well at the moment, its flexibility leads to the personal opinion of developers directing information flow rather than a defined framework, ultimately resulting in difficult-to-maintain code as only those who have already contributed fully understand the operations taking place.

Thus, this PR and associated branch aims to recreate the current functionality on the master branch using Django as a backend Python web framework, replacing Flask. To ease the transition the template engine is kept as Jinja2.

Before ready to merge into development, the following tasks must be completed:
- [X] Configure development server to serve static files
- [X] Jinja2 setup for Django
- [X] Model descriptions for `EmailTemplate`s and other objects in `models.py`
- [X] Describing user inputs as Django Forms in `forms.py`
- [x] Re-formatting logic for filling out unfilled `EmailTemplate` objects
- [x] Re-formatting logic for parsing user-submitted email templates
- [x] `/postcode/` endpoint for Javascript lookup of addresses 
- [ ] Storing existing templates in Django database
- [ ] Configuring Heroku `development` deployment for a Django application, and investigating changes needed to `master` Heroku deployment

Additionally, some information flows could be streamlined during this re-write:
- [x] Rewrite landing page template so that the form is created dynamically from the Django Form object
- [x] Perform a single lookup of the user's postcode to get both addresses and constituency (if an API is available)
- [x] Investigate a single template-filling stage,  is there a benefit to using  `render()` for the email bodies themselves?

fpervaiz committed:
Boilerplate code and UI

fpervaiz committed:
Boilerplate code and UI

behradkoohy committed:
Merge branch 'template-submission-ui' of https://github.com/Write-For-Change/emaildrafter into feature/db-admin-ui

behradkoohy committed:
added submission form in

behradkoohy committed:
added submission form in

behradkoohy committed:
removed double click bug in submission

davidswarbrick committed:
Merge single template changes (#45)

* Added ability to access single template

* initial work on single template page

* Added fetch template by name function

* Succesfully find template by slug

* Added slugify to requirements + find function

* Removed myDb reference

davidswarbrick committed:
Merge single template changes (#45)

* Added ability to access single template

* initial work on single template page

* Added fetch template by name function

* Succesfully find template by slug

* Added slugify to requirements + find function

* Removed myDb reference

behradkoohy committed:
try catch

davidswarbrick committed:
Template submitter class for storage in db

davidswarbrick committed:
Formulated Template submission syntax and substitution process, fleshed out Form object to include more attributes of EmailTemplate objects, and written tests for new regex implementations

davidswarbrick committed:
Updated README

davidswarbrick committed:
Revert "Updated README"

This reverts commit 3863bd5cadf0c58d699aa2870c3319782ffdecba.
Attempting to only change readme once

davidswarbrick committed:
Updated README to include template information

davidswarbrick committed:
Updated tests to reflect choices about mis-matched cases in field names.

davidswarbrick committed:
Adjusted validation to use set notation for clarity, added boilerplate for form object filling & creation method

davidswarbrick committed:
First test on local dev was successful

davidswarbrick committed:
Initial commit of Django project setup, using a Django app for emaildrafter (so that REST apps etc can be in their own directories)

davidswarbrick committed:
Moved emaildrafter to project root as that is more standard for Django. Replaced secret key with generated one for development environment - Important to note this branch is not production ready at all, and settings.py must be altered before deploying in any setting.

davidswarbrick committed:
Model relations set up

davidswarbrick committed:
Serving static files correctly, implemented homepage and about us, need to address postcode lookup (but might want to embed into Django Form)

tjgurwara99 commented on emaildrafter/admin.py:
```diff
@@ -0,0 +1,3 @@
+from django.contrib import admin
+
+# Register your models here.
```
Maybe add your model.py's models ModelAdmin here to get admin access CRUD on your models.

tjgurwara99 commented on emaildrafter/views.py:
```diff
@@ -0,0 +1,19 @@
+
+def index(request):
+    if request.method == "POST":
+        form = UserForm(request.POST)
```
This is good. Is there a reason for using function based view instead of class based?

davidswarbrick replied:
I've started with function based views rather than class based as this emulates the Flask app behaviour so should ease the everyone's transition to the Django way of doing things.

tjgurwara99 replied:
Ah okay, makes sense!

davidswarbrick committed:
Added some docstrings and consideration of how to fill template bodies in a django-friendly way.

davidswarbrick committed:
Removed extraneous definition of template submission form

davidswarbrick commented:
The force push was to rebase this branch on `feature/db-admin-ui` so that I could replicate that functionality in the Django swap/upgrade. Tests must be reconfigured for the Django layout.

davidswarbrick committed:
Removed HEAD in requirements.txt leftover from rebase

davidswarbrick committed:
Added logic for filling and validating template bodies, plus generating slugs from template names.

davidswarbrick committed:
Moved address lookup to an external apis file, and added constituency to user form, returned from postcode api endpoint

tjgurwara99 commented on writeforchange/settings.py:
```diff
@@ -0,0 +1,129 @@
+# Database
+# https://docs.djangoproject.com/en/3.0/ref/settings/#databases
+
+DATABASES = {
```
Sorry I missed this, but should we not change to a different database than sqlite3 unless the speed is not an issue for the api calls.

davidswarbrick replied:
Yeah we can do, I have no previous database experience so happy to hear any opinions on this. @behradkoohy is a fan of PostgreSQL right?

tjgurwara99 replied:
Yeah, PostgreSQL is good too! Do you guys think there would be read heavy api's in this project though? That would be the only reason to not use PostgreSQL over MySQL. Otherwise, I don't see a reason not to use it.

In most cases, we can also implement both for different cases like so:

```python
DATABASES = {
    'default': {
        'NAME': 'main_data',
        'ENGINE': 'django.db.backends.postgresql',
        'USER': 'postgres_user',
        'PASSWORD': 'abcde'
    },
    'users': {
        'NAME': 'user_data',
        'ENGINE': 'django.db.backends.mysql',
        'USER': 'mysql_user',
        'PASSWORD': 'abcde'
    }
}
```

but I don't think this application/project warrants such an implementation.

davidswarbrick committed:
User input now Django Form, with JS alterations

 - Rending form automatically from Django Form
   (CSS adjustments still required)

 - Getting postcode data from /postcode/ endpoint

 - Filling address + constituency from JSON response

 - Form validating successfully

davidswarbrick committed:
Added models to admin, tested template filling

 - Added all the relevant models to the admin UI and submitted an example template and MP

 - Succesfully filled the template with UserForm and Target data

 - Adjusted the target property of a template to correspond to either the specific target or MP

 - ToDo: Remove the blank space in the Email Body textarea widget, adjust landing page CSS, switch to PostgreSQL db and re-implement topic and single template endpoints.

davidswarbrick commented:
The Django implementation of emaildrafter is now working:

1. A user is asked for their name and postcode in a similar form as the Flask implementation, but now it is automatically generated by Django's Form API thus some CSS hooks are missing/broken.
2. The `/postcode/` endpoint now returns both the addresses and constituency of the postcode, which are inserted into relevant fields in the form, initially hidden to the user by the Javacript function.
3. The form is validated on submission, and then passed to a `draft_templates()` function as previously, however the templates are now stored and accessed as Django Model objects. The filling function either uses the specific target of the template, stored in a separate database table, or an MP which is looked up in its own database table using the constituency of the user as submitted in the initial form.
4. Template bodies are stored as `TextField`s in the database. To fill them, an `EmailBody` object is intialised (defined in `emailbody.py`), which extends the Python Template class to implement `%TONAME`, `%YOURNAME`, etc. shortcuts to information about the target and user. Also defined in `emailbody.py` is a validator to check submitted template bodies only contain fields which can be filled by this app. Between `EmailTemplate` in `models.py` and `EmailBody` in `emailbody.py` most of the functionality of `emailtemplates.py` is recreated, although there are some extra validation steps which may be useful and have not yet been recreated.
5. Filled templates are returned and displayed in `results.html` as before.

Updated To Do:
- [ ] Adjust CSS on landing page to account for different form definition
- [ ] Remove leading whitespace in text area widget for rendered email bodies in `results.html`
- [ ] Select only templates that are `public` and `admin_approved` for display on the homepage.
- [ ] Switch database to PostgreSQL - speed improvement future-proofs the project against REST APIs or scaling, other developers have shown a preference. MySQL has been used so far for ease of testing locally, but once a database has been deployed to staging or production I suspect the complexity increase will be equivalent for any choice of DB, thus it makes sense to go with a familiar option to other members of the team.
- [ ] Recreate `/topic/` and `/template/` endpoints
- [ ] Remove Flask implementation code (currently still present on the branch as a reference).

Carried forward:
- [ ] Store existing templates in Django database
- [ ] Configure Heroku `development` deployment for a Django application, and investigating changes needed to master Heroku deployment

Identified extensions:
- [ ] Template submission from non-admins - Django's `forms.ModelForm` should ease this considerably.
- [ ] Moderator accounts with differing levels of access to the administrator interface.
- [ ] Selection box for user of the templates they wish to be filled out - could add a `forms.ChoiceField()` with the choices set to a query of `EmailTemplate.objects()`
- [ ] Responsive CSS, for example on the box surrounding user input.

Commit:
edited Procfile, added settings.py

Commit:
added settings.py whoops
